crates/ruff_source_file/src/locator.rs

```diff
@@ -388,6 +388,23 @@ impl<'a> Locator<'a> {
         &self.contents[usize::from(offset)..]
     }
 
+    /// Finds the closest [`TextSize`] not exceeding the offset for which `is_char_boundary` is
+    /// `true`.
+    ///
+    /// Can be replaced with `str#floor_char_boundary` once it's stable.
+    pub fn floor_char_boundary(&self, offset: TextSize) -> TextSize {
+        if offset >= self.text_len() {
+            self.text_len()
+        } else {
+            // We know that the character boundary is within four bytes.
+            (0u32..=3u32)
+                .map(TextSize::from)
+                .filter_map(|index| offset.checked_sub(index))
+                .find(|offset| self.contents.is_char_boundary(offset.to_usize()))
+                .unwrap_or_default()
+        }
+    }
```

Comment:
I moved this into a dedicated method on `Locator`, based on the `str#floor_char_boundary` API which is unstable (so we can't use it) but has this rough pattern.

Reply:
Thank you for advice!